From fc967f0200160ca234c6bc40e41d7962adb6fd8f Mon Sep 17 00:00:00 2001
From: Vadik Likholetov <vadikas@gmail.com>
Date: Tue, 18 Aug 2026 12:23:32 +0300
Subject: [PATCH] iommu: tegra-pkvm: diagnose translated DMA

Keep the Android donor execute-never policy, add the missing non-cacheable
mapping attribute, and compare software leaves against hardware ATOS results
on both Tegra SMMU instances.
---
 arch/arm64/kvm/hyp/nvhe/iommu/tegra-smmu.c | 27 ++++++++++++++++++++--
 drivers/iommu/arm-smmu-tegra-pkvm.c        | 13 ++++++++---
 include/kvm/tegra-smmu-pkvm.h              |  5 ++++
 3 files changed, 40 insertions(+), 5 deletions(-)

diff --git a/arch/arm64/kvm/hyp/nvhe/iommu/tegra-smmu.c b/arch/arm64/kvm/hyp/nvhe/iommu/tegra-smmu.c
index 01d3d5de2..b75f4e05e 100644
--- a/arch/arm64/kvm/hyp/nvhe/iommu/tegra-smmu.c
+++ b/arch/arm64/kvm/hyp/nvhe/iommu/tegra-smmu.c
@@ -27,6 +27,7 @@
 
 #define PKVM_TEGRA_TLB_SPINS		1000000
 #define PKVM_TEGRA_MGBE_RESET_SPINS	2000000
+#define PKVM_TEGRA_ATS_SPINS		100000
 
 /* Temporary boot diagnostics for the nvidia-jetson-orin-agx-pkvm-debug target. */
 #define PKVM_TEGRA_DIAG_INIT_PARAMS_DONATE	(-1001)
@@ -717,10 +718,12 @@ static int tegra_map_pages(struct kvm_hyp_iommu_domain *core_domain,
 		pgt_prot |= KVM_PGTABLE_PROT_R;
 	if (prot & IOMMU_WRITE)
 		pgt_prot |= KVM_PGTABLE_PROT_W;
-	if (prot & IOMMU_MMIO)
-		pgt_prot |= KVM_PGTABLE_PROT_DEVICE;
 	if (!pgt_prot)
 		return -EINVAL;
+	if (prot & IOMMU_MMIO)
+		pgt_prot |= KVM_PGTABLE_PROT_DEVICE;
+	else if (!(prot & IOMMU_CACHE))
+		pgt_prot |= KVM_PGTABLE_PROT_NORMAL_NC;
 	domain->debug_iova = iova;
 
 	ret = iommu_pkvm_use_dma(paddr, size);
@@ -891,6 +894,8 @@ static int tegra_debug_read(pkvm_handle_t iommu_id, u32 selector,
 		u32 op = FIELD_GET(PKVM_TEGRA_DEBUG_DOMAIN_OP, selector);
 		kvm_pte_t pte;
 		s8 level;
+		unsigned int instance;
+		unsigned int spin;
 
 		if (domain_id >= KVM_IOMMU_MAX_HOST_DOMAINS)
 			return -EINVAL;
@@ -933,6 +938,24 @@ static int tegra_debug_read(pkvm_handle_t iommu_id, u32 selector,
 				  readl_relaxed(cb + PKVM_SMMU_CB_FSYNR0);
 			*value1 = readq_relaxed(cb + PKVM_SMMU_CB_FAR);
 			return 0;
+		case 5:
+		case 6:
+			instance = op - 5;
+			if (instance >= smmu->params->num_instances)
+				return -ENOENT;
+			cb = tegra_smmu_cb(smmu, instance, domain->cb);
+			writeq_relaxed(domain->debug_iova & PAGE_MASK,
+				       cb + PKVM_SMMU_CB_ATS1PR);
+			for (spin = 0; spin < PKVM_TEGRA_ATS_SPINS; spin++) {
+				if (!(readl_relaxed(cb + PKVM_SMMU_CB_ATSR) &
+				      PKVM_SMMU_CB_ATSR_ACTIVE)) {
+					*value0 = readq_relaxed(cb + PKVM_SMMU_CB_PAR);
+					*value1 = domain->debug_iova;
+					return 0;
+				}
+				cpu_relax();
+			}
+			return -ETIMEDOUT;
 		default:
 			return -EINVAL;
 		}
diff --git a/drivers/iommu/arm-smmu-tegra-pkvm.c b/drivers/iommu/arm-smmu-tegra-pkvm.c
index 269890463..b08336bce 100644
--- a/drivers/iommu/arm-smmu-tegra-pkvm.c
+++ b/drivers/iommu/arm-smmu-tegra-pkvm.c
@@ -248,14 +248,21 @@ static int pkvm_tegra_map_pages(struct iommu_domain *domain,
 				gfp_t gfp, size_t *mapped)
 {
 	struct pkvm_tegra_domain *tegra_domain = to_pkvm_tegra_domain(domain);
-	u64 value[5][2] = { };
-	int debug_ret[5];
+	u64 value[7][2] = { };
+	int debug_ret[7];
+	bool debug;
 	int ret;
 	int op;
 
+	debug = tegra_domain->debug_maps++ < 32;
+	if (debug)
+		pr_err("tegra-pkvm-map-begin: smmu=%llu domain=%llu iova=%#lx pa=%pa pgsize=%zu pgcount=%zu prot=%#x\n",
+		       (unsigned long long)tegra_domain->smmu->id,
+		       (unsigned long long)tegra_domain->id, iova, &paddr,
+		       pgsize, pgcount, prot);
 	ret = kvm_iommu_map_pages(tegra_domain->id, iova, paddr, pgsize,
 				  pgcount, prot, gfp, mapped);
-	if (tegra_domain->debug_maps++ >= 8)
+	if (!debug)
 		return ret;
 
 	for (op = 0; op < ARRAY_SIZE(debug_ret); op++)
diff --git a/include/kvm/tegra-smmu-pkvm.h b/include/kvm/tegra-smmu-pkvm.h
index ed26baa7c..3a0fb200d 100644
--- a/include/kvm/tegra-smmu-pkvm.h
+++ b/include/kvm/tegra-smmu-pkvm.h
@@ -100,6 +100,8 @@ struct pkvm_tegra_smmu_device {
 #define PKVM_SMMU_SCTLR_M		BIT(0)
 #define PKVM_SMMU_CB_TTBR0		0x020
 #define PKVM_SMMU_CB_TCR		0x030
+#define PKVM_SMMU_CB_PAR		0x050
+#define PKVM_SMMU_CB_PAR_F		BIT(0)
 #define PKVM_SMMU_VTCR_RES1		BIT(31)
 #define PKVM_SMMU_VTCR_PS		GENMASK(18, 16)
 #define PKVM_SMMU_VTCR_SH0		GENMASK(13, 12)
@@ -110,6 +112,9 @@ struct pkvm_tegra_smmu_device {
 #define PKVM_SMMU_CB_FSR		0x058
 #define PKVM_SMMU_CB_FAR		0x060
 #define PKVM_SMMU_CB_FSYNR0		0x068
+#define PKVM_SMMU_CB_ATS1PR		0x800
+#define PKVM_SMMU_CB_ATSR		0x8f0
+#define PKVM_SMMU_CB_ATSR_ACTIVE	BIT(0)
 #define PKVM_SMMU_CB_TLBIIPAS2		0x630
 #define PKVM_SMMU_CB_TLBIIPAS2L	0x638
 #define PKVM_SMMU_CB_TLBSYNC		0x7f0
-- 
2.55.0
